### ISSUE TITLE
u32 interpreted as i32 when passed to JS

### ISSUE THREAD

**Issue by Procrat** [rustwasm/wasm-bindgen#1388]
### Describe the Bug

A valid u32 larger than _2^31_ turns negative when passed to JavaScript.

### Steps to Reproduce

1. Take the `import_js` example.
2. Change the `x.set_number(10)` line to `x.set_number(2147483647)` or any other u32 larger than _2^31_.
3. See the negative number in the console.

### Expected Behavior

I expect to see a positive number.

### Actual Behavior

We see a negative number.

### Additional Context

Tested on wasm-bindgen 0.2.40.

**Comment by Pauan** [comment]
I believe this can be fixed by having wasm-bindgen run `number >>> 0` on the JS receiving side for `u32`.

We should also test to make sure that it works correctly when sending `u32` *to* Rust.

**Comment by Procrat** [comment]
When you put it like that, this is perhaps the wrong repo to raise this issue. If the generated wasm is already wrong, it's probably not wasm-bindgen's job to patch that up, right? The fault happens in rustc, right?

Just to confirm that that is what's actually happening, I compiled
```rust
#[no_mangle]
pub extern "C" fn f() -> u32 {
    2u32.pow(31)
}
```
and made it readable with `wasm2wat`:
```wat
(module
  (type (;0;) (func))
  (type (;1;) (func (result i32)))
  (func $__wasm_call_ctors (type 0))
  (func $f (type 1) (result i32)
    i32.const -2147483648)
  (table (;0;) 1 1 anyfunc)
  (memory (;0;) 16)
  (global (;0;) (mut i32) (i32.const 1048576))
  (global (;1;) i32 (i32.const 1048576))
  (global (;2;) i32 (i32.const 1048576))
  (export "memory" (memory 0))
  (export "__heap_base" (global 1))
  (export "__data_end" (global 2))
  (export "f" (func $f)))
```
So it's already negative there.

Where should I report this?

**Comment by Pauan** [comment]
> If the generated wasm is already wrong, it's probably not wasm-bindgen's job to patch that up, right?

Wasm does not support `u32`, it only supports `i32`. So this cannot be fixed in Rust.

Instead, the `.js` snippets need to do a `number >>> 0` cast (to convert from `i32` to `u32` on the JS side).

wasm-bindgen generates the `.js` snippets, and therefore this is definitely wasm-bindgen's responsibility.

**Comment by Procrat** [comment]
I'm a bit surprised by this fix. Even though u32 isn't supported in Wasm, it can be represented with an i64.

I'm I then right in concluding that the rustwasm team makes the conscious decision that rustc's output is only an intermediary (and is thus deemed fine to be wrong by itself), and that for any working Rust to Wasm compilation, the use `wasm-bindgen` and the use of JavaScript is required? Or am I missing something?

**Comment by Pauan** [comment]
@Procrat JavaScript does not have support for `i64` (it throws an exception if you try to use it).

Using `f64` works, but that doubles the memory usage and has a performance cost for conversion. There's no reason to use `f64`, since `i32` works perfectly (it has the same bit width and semantics as `u32`).

> I'm I then right in concluding that the rustwasm team makes the conscious decision that rustc's output is only an intermediary (and is thus deemed fine to be wrong by itself)

It isn't "wrong", it's simply following the Wasm spec as closely as possible: Wasm doesn't support `u32`, but an `i32` works perfectly fine (because the bit representation is identical), so that's what Rust uses.

Normally this isn't a problem, since Rust can just always use `i32` and it will have the correct semantics. It is only when communicating with *other* languages (like JavaScript) where this becomes a problem.

> and that for any working Rust to Wasm compilation, the use wasm-bindgen and the use of JavaScript is required?

Yes, if you want to communicate with JS you will need an extra layer (such as `wasm-bindgen`) to handle the data exchange back and forth from Rust <-> JS.

This data exchange is quite complex: it involves type conversions, and also creating a stack/heap on the JS side to hold JS objects (since wasm currently cannot send objects back and forth, it can only send numbers, so `wasm-bindgen` must store the JS objects in the JS heap and then send a numeric index to Rust).

As the `wasm32-unknown-unknown` name implies, the target is "unknown", which means that Rust cannot make any assumptions about the host environment: it might be a browser, or Node, or a C# Wasm interpreter, or a blockchain, or...

Basically, the `wasm32-unknown-unknown` means "all bets are off, you're on your own".

There have been discussions about adding other targets, like `wasm32-unknown-web` or the more recent `wasm32-unknown-wasi`, but for now we're using `wasm32-unknown-unknown`, because right now it's the only thing Rust supports.

**Comment by Pauan** [comment]
By the way, Wasm will soon have support for `anyref`, which removes the need to have a JS stack/heap, and eventually when host bindings are implemented Wasm will [also have support for sending `u32` values to JS](https://github.com/WebAssembly/host-bindings/blob/0ab2d84f35f9b7062693d853aea6055a31defec4/proposals/host-bindings/Overview.md#import-argument-binding-types).

It will still be needed to use `wasm-bindgen` for various other reasons (including the ability to use npm packages and `.js` files).

All of this stuff is just internal implementation details that you shouldn't be worrying about (unless you're just curious and are interested in learning!)

**Comment by Pauan** [comment]
You might be wondering why we couldn't just take `wasm-bindgen` and put it into `rustc`.

We *could* do that, but any changes to `rustc` require pretty extensive review and agreement from the core contributors. It's a very slow process (for good reason!)

So it's a *lot* easier to have `wasm-bindgen` be a separate tool which can be rapidly changed/improved at its own pace.

Maybe eventually there can be stronger integration with Rust itself, and maybe eventually `wasm-bindgen` won't be needed anymore, but that's all far in the future.

**Comment by Procrat** [comment]
Thanks for taking the time to write such a thorough explanation!

I'm thinking about an issue of separation of concerns here. `rustc` just takes care of the Rust -> Wasm compilation without any knowledge of the embedder, which might not have a JavaScript VM. It's only by putting `wasm-bindgen` into the picture that we have Rust <-> JavaScript interop, right? So, specifically thinking of a future case where we want to call Wasm from a non-JS embedding, without any wrapper like wasm-bindgen, you can only talk in i32/i64/f32/f64s to that Wasm module. I'd say it's then a programmer mistake if a plain Rust function (still without wasm-bindgen!) accepts anything other than i32/i64/f32/f64s and I think I would expect a compiler error. If we on the other hand _do_ have a JS embedding and want to pass through other types between Rust and JS, we can start introducing wasm-bindgen to the code, and then it would be wasm-bindgen's responsibility to the type conversions on both ends of the FFI.

As an example, I guess my main surprise is that
```rust
#[no_mangle]
pub extern "C" fn f() -> u32 {
    2u32.pow(31)
}
```
is allowed to compile, since, like you said, a Wasm module can't return a u32.

**Comment by Pauan** [comment]
@Procrat I agree that is rather surprising.

I'm not 100% sure, but I think that it's because it's not actually exporting the function to wasm... it's exporting the function to C (hence the `extern "C"` part). And C *does* have `u32`, so Rust accepts it just fine.

Essentially, there really *should* be an `extern "wasm"`, but there isn't, so we currently abuse `extern "C"` a bit.

**Comment by Procrat** [comment]
Ah, I see. That makes sense! Thanks for the explanation!

**Comment by AhmetKaanGuney** [comment]
This discussion was very eye opening thank you all :)
